server/conn_stmt_test.go

```diff
@@ -352,3 +357,269 @@ func TestCursorFetchErrorInFetch(t *testing.T) {
+		mysql.CursorTypeReadOnly, 0x1, 0x0, 0x0, 0x0,
+		0x0, 0x1, 0x3, 0x0,
+		0x1, 0x0, 0x0, 0x0,
+	)))
+
+	out := c.GetOutput()
+	// fetch one row
+	expected := expectedLonglongFetchResult(t, c, 1)
+	require.NoError(t, c.Dispatch(ctx, appendUint32(appendUint32([]byte{mysql.ComStmtFetch}, uint32(stmt.ID())), 1)))
+	require.NoError(t, c.flush(context.Background()))
+	require.Equal(t, expected, out.Bytes())
+	// reset the statement
+	require.NoError(t, c.Dispatch(ctx, append(
+		appendUint32([]byte{mysql.ComStmtReset}, uint32(stmt.ID())),
+	)))
+	// the following fetch will fail
+	require.Error(t, c.Dispatch(ctx, appendUint32(appendUint32([]byte{mysql.ComStmtFetch}, uint32(stmt.ID())), 1)))
+}
+
+func dispatchSendLongData(c *mockConn, stmtID int, paramIndex uint16, parameter []byte) error {
+	appendUint16 := binary.LittleEndian.AppendUint16
+	appendUint32 := binary.LittleEndian.AppendUint32
+
+	return c.Dispatch(context.Background(),
+		append(
+			appendUint16(
+				appendUint32([]byte{mysql.ComStmtSendLongData}, uint32(stmtID)),
+				paramIndex), // the index of parameter
+			parameter..., // the parameter
+		),
+	)
+}
+
+func TestCursorFetchSendLongData(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	srv := CreateMockServer(t, store)
+	srv.SetDomain(dom)
+	defer srv.Close()
+
+	appendUint32 := binary.LittleEndian.AppendUint32
+	appendUint64 := binary.LittleEndian.AppendUint64
+
+	ctx := context.Background()
+	c := CreateMockConn(t, srv).(*mockConn)
+	c.capability = mysql.ClientProtocol41
+
+	tk := testkit.NewTestKitWithSession(t, store, c.Context().Session)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(id_1 BIGINT, id_2 BIGINT)")
+	tk.MustExec("insert into t values (1, 1), (1, 2)")
+
+	// as the `ComStmtSendLongData` only sends bytes parameters, use `SUBSTR(HEX(?), 1, 2)` to convert a little endian 1 to
+	// a string "1"
+	stmt, _, _, err := c.Context().Prepare("select id_1 from t where id_1 = cast(SUBSTR(HEX(?), 1, 2) as UNSIGNED)")
+	require.NoError(t, err)
+
+	// send a parameter to the server
+	require.NoError(t, dispatchSendLongData(c, stmt.ID(), 0, appendUint64([]byte{}, 1)))
+
+	// execute the statement without argument
+	require.NoError(t, c.Dispatch(ctx, append(
+		appendUint32([]byte{mysql.ComStmtExecute}, uint32(stmt.ID())),
+		mysql.CursorTypeReadOnly, 0x1, 0x0, 0x0, 0x0,
+		0x0, 0x0,
+	)))
+
+	out := c.GetOutput()
+	// fetch one row
+	expected := expectedLonglongFetchResult(t, c, 1)
+	require.NoError(t, c.Dispatch(ctx, appendUint32(appendUint32([]byte{mysql.ComStmtFetch}, uint32(stmt.ID())), 1)))
+	require.NoError(t, c.flush(context.Background()))
+	require.Equal(t, expected, out.Bytes())
+}
+
+func TestCursorFetchSendLongDataReset(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	srv := CreateMockServer(t, store)
+	srv.SetDomain(dom)
+	defer srv.Close()
+
+	appendUint32 := binary.LittleEndian.AppendUint32
```

Comment:
How about change `appendUint16`, `appendUint32`, `appendUint64` funcs to global variables?